Mathlib/Topology/Constructible.lean

```diff
@@ -0,0 +1,303 @@
+    rw [union_inter_distrib_right, ← image_inter_preimage]
+    exact ((hUcomp (hf.isCompact_preimage hVcomp) (hVopen.preimage hf.continuous)).image
+      hf.continuous).union <| hfcomp hVcomp hVopen
+  | union s hs t ht hs' ht' => rw [image_union]; exact hs'.union ht'
+  | compl s hs hs' =>
+    rw [← range_diff_image hf.injective]
+    exact (hfcomp.isConstructible hf.isClosed_range.isOpen_compl).of_compl.sdiff hs'
+
+section CompactSpace
+variable [CompactSpace α] {P : ∀ s : Set α, IsConstructible s → Prop} {b : ι → Set α}
+
+lemma IsRetroCompact.isCompact (hs : IsRetroCompact s) : IsCompact s := by
+  simpa using hs CompactSpace.isCompact_univ
+
+lemma TopologicalSpace.IsTopologicalBasis.isRetroCompact_iff_isCompact
+    (basis : IsTopologicalBasis (range b)) (compact_inter : ∀ i j, IsCompact (b i ∩ b j))
+    (hU : IsOpen U) : IsRetroCompact U ↔ IsCompact U := by
+  refine ⟨IsRetroCompact.isCompact, fun hU' {V} hV' hV ↦ ?_⟩
+  have hb (i : PLift ι) : IsCompact (b i.down) := by simpa using compact_inter i.down i.down
+  have := isCompact_open_iff_eq_finite_iUnion_of_isTopologicalBasis (b ∘ PLift.down)
+    (by simpa [PLift.down_surjective.range_comp] using basis) hb
+  obtain ⟨s, hs, rfl⟩ := (this U).mp ⟨hU', hU⟩
+  obtain ⟨t, ht, rfl⟩ := (this V).mp ⟨hV', hV⟩
+  simp only [Set.iUnion_inter, Set.inter_iUnion]
+  exact ht.isCompact_biUnion fun _ _ ↦ hs.isCompact_biUnion fun _ _ ↦ compact_inter ..
+
+lemma TopologicalSpace.IsTopologicalBasis.isRetroCompact (basis : IsTopologicalBasis (range b))
+    (compact_inter : ∀ i j, IsCompact (b i ∩ b j)) (i : ι) : IsRetroCompact (b i) :=
+  (basis.isRetroCompact_iff_isCompact compact_inter <| basis.isOpen <| mem_range_self _).2 <| by
+    simpa using compact_inter i i
+
+lemma TopologicalSpace.IsTopologicalBasis.isConstructible (basis : IsTopologicalBasis (range b))
+    (compact_inter : ∀ i j, IsCompact (b i ∩ b j)) (i : ι) : IsConstructible (b i) :=
+  (basis.isRetroCompact compact_inter _).isConstructible <| basis.isOpen <| mem_range_self _
+
+proof_wanted IsConstructible.induction_of_isTopologicalBasis {ι : Type*} (b : ι → Set α)
+    (basis : IsTopologicalBasis (range b)) (compact_inter : ∀ i j, IsCompact (b i ∩ b j))
+    (sdiff : ∀ i s (hs : Set.Finite s), P (b i \ ⋃ j ∈ s, b j)
+      ((basis.isConstructible compact_inter _).sdiff <| .biUnion hs fun i _ ↦
+        basis.isConstructible compact_inter _))
+    (union : ∀ s hs t ht, P s hs → P t ht → P (s ∪ t) (hs.union ht))
+    (s : Set α) (hs : IsConstructible s) : P s hs
+
+end CompactSpace
+
+/-! ### Locally constructible sets -/
+
+/-- A set in a topological space is locally constructible, if it can be partitioned along a finite
+open cover such that every part is constructible -/
+@[stacks 005G]
+def IsLocallyConstructible (s : Set α) : Prop :=
+  ∃ F : Set (Set α), (∀ x, ∃ U ∈ F, x ∈ U) ∧ (∀ U ∈ F, IsOpen U) ∧ ∀ U ∈ F, IsConstructible (s ∩ U)
+
+lemma IsLocallyConstructible.of_iUnion {ι : Sort*} {f : ι → Set α} (cover : ∀ x, ∃ i, x ∈ f i)
```

Comment:
I'm surprised this lemma cannot be generalized to `IsConstructible s`.

Reply:
This is just a restatement of `IsLocallyConstructible` in terms of `iUnion`. `ι` can be infinite